### PR TITLE
ArC: implement equals/hashCode for TypeVariableImpl and WildcardTypeImpl

### PR DESCRIPTION
**Pull request by mkouba**

### CODE REVIEW

**Review by Ladicek — CHANGES_REQUESTED** [review]
The hash functions here are not consistent with the JDK `TypeVariable` and `WildcardType` implementations. I know these are not a public contract, but it's the only sensible implementation, otherwise a type obtained e.g. from a `TypeLiteral` won't be usable. Guava for example does the same thing, see https://github.com/google/guava/blob/master/guava/src/com/google/common/reflect/Types.java

**Comment by Ladicek** [comment]
Also, `equals` shouldn't assume that the other object is of the same class. Our `WildcardTypeImpl` must be equal to the corresponding wildcard type obtained e.g. again from `TypeLiteral`.

**Comment by mkouba** [comment]
> Also, `equals` shouldn't assume that the other object is of the same class. Our `WildcardTypeImpl` must be equal to the corresponding wildcard type obtained e.g. again from `TypeLiteral`.

I'm not sure I'm following. We don't test that the `WildcardTypeImpl`  is of the same class but it must implement `WildcardType`... which is what Guava does as well.

**Comment by Ladicek** [comment]
OK, my bad, the `equals` methods are fine. I probably got confused by the tests that only use our implementations of types. We could use `TypeLiteral` to obtain JDK implementations of types to test, for extra bonus points :-)

**Comment by mkouba** [comment]
@Ladicek There's a problem with `io.quarkus.arc.impl.TypeVariableImpl.hashCode()` though - JDK/guava do use `genericDeclaration.hashCode() ^ name.hashCode()` but we don't support `genericDeclaration` at all. The same applies to `equals()`, bounds are ignored but the `name` and `genericDeclaration` are considered...

File: independent-projects/arc/runtime/src/test/java/io/quarkus/arc/impl/ParameterizedTypeImplTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+        // List<? extends Number>
+        ParameterizedTypeImpl parameterizedType3 = new ParameterizedTypeImpl(List.class,
+                WildcardTypeImpl.withUpperBound(Number.class));
+        assertEquals(parameterizedType3,
+                new TypeLiteral<List<? extends Number>>() {
+                }.getType());
+        assertEquals(parameterizedType3.hashCode(), new ParameterizedTypeImpl(List.class,
+                WildcardTypeImpl.withUpperBound(Number.class)).hashCode());
+        assertNotEquals(parameterizedType3, parameterizedType1);
```
Comment (Ladicek) [inline independent-projects/arc/runtime/src/test/java/io/quarkus/arc/impl/ParameterizedTypeImplTest.java:41]: Hash code for the 2nd and 3rd case, similarly to the 1st case?

**Review by Ladicek — APPROVED** [review]

**Review by manovotn — APPROVED** [review]

**Review by Ladicek — APPROVED** [review]

**Pull request merged**